api/client.go

```diff
@@ -25,6 +25,36 @@ import (
 	"time"
 )
 
+func NewErrorAPI(err error, warnings []string) Error {
+	if err == nil && warnings == nil {
+		return nil
+	}
+	return &ErrorAPI{err, warnings}
+}
+
+type ErrorAPI struct {
+	err      error
+	warnings []string
+}
+
+func (w *ErrorAPI) Err() error {
+	return w.err
+}
+
+func (w *ErrorAPI) Error() string {
+	return w.err.Error()
+}
+
+func (w *ErrorAPI) Warnings() []string {
+	return w.warnings
+}
+
+type Error interface {
+	error
+	Err() error
```

Comment:
Why the need for Err() and Error()?

Reply:
`Err()` returns the error itself. This is important if you want to check if the error is something in particular. This is especially important as we have been returning the err directly until now. So old code that did something like:

```
ret, err := client.DoSomething()
if err != nil {
    return nil, err
}
``

would now be able to do something like
```
ret, err := client.DoSomething()
if err != nil && err.Err() != nil {
    return nil, err.Err()
}
``

Reply:
hm I still don't see the need for this.
Can you give an example for how and where it is needed and why it can't be achieved by just using err.Error() ?

Reply:
err.Error() returns a string, if you need to check that is a specific error type or something then it is not sufficient.

For example, when we do a client get we could get back an error from the HTTP library. In the client it looks like so:
```
	_, body, apiErr := h.client.Do(ctx, req)
	if apiErr != nil {
		return AlertsResult{}, apiErr
	}
```

So if I wanted to handle the specific error type of `http.ProtocolError` I'd need to do something like:

```
if err != nil {
    if typedErr, ok := err.Err().(*http.ProtocolError); ok {
        // do something else
    }
    return nil, err
}
```

If we *only* have the string representation the people are forced to resort to string matching, which is a bad plan :)

Reply:
Yes I agree. Lets keep just 

```
// Error .....
type Error interface {
	// Error ......
	Error() error
	// Warnings returns a list of warnings
	Warnings() []string
}
```

There are still few places without comments the go-lint complains about it.

Reply:
So you want to only have those 2 methods? This would mean the return type doesn't adhere to the error interface anymore meaning this is a completely breaking change.

 If so I'd call the errror one err (error is for string, err is for an actual error type).

Reply:
As I'm thinking about this more, I think we want to stick with:

```
type Error interface{
    error
    Err() error
    Warnings() []string
}
```

and change the implementations of `Error()` (the string return) to check the warnings. This way its more-or-less backwards compatible -- although it would mean that warnings are errors (if you wanted to skip them you'd have to check the Err() method). This would make it more compatible, and a little less clunky.

If we want the default behavior to be that people have to check if it was an error or warning then I think going to some interface that doesn't adhere to the `error` interface would be okay -- it will just be completely incompatible with all client usage.

Reply:
Yes you are right lets keep as is for now and will refactor if needed.

Reply:
@krasi-georgiev to clarify, do you want me to change the `Error()` implementation to return something if there was a warning (e.g. if no error but warnings, return the warnings concated somehow)?

Reply:
@krasi-georgiev bump :)

Reply:
Sorry, I am at kubecon so will be slow on replies.
I meant to keep it as
```
type Error interface{
    error
    Err() error
    Warnings() []string
}
```